tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2156,16 +2201,16 @@ static bool CombineBasicTypes(ArBasicKind LeftKind,
                               ArBasicKind RightKind,
                               _Out_ ArBasicKind* pOutKind)
 {
-  if ((LeftKind < 0 || LeftKind >= AR_BASIC_COUNT) ||
-    (RightKind < 0 || RightKind >= AR_BASIC_COUNT)) {
-    return false;
-  }
-
   if (LeftKind == RightKind) {
     *pOutKind = LeftKind;
     return true;
   }
 
+  if ((LeftKind < 0 || LeftKind >= AR_BASIC_COUNT) ||
```

Comment:
I think the point here was that this function should only be used to combine "basic" types, not object types.  If we are using this for object types, I think there could be a problem.

For instance, I fear there may not be sufficient checks to ensure that the object types of the two inputs are identical, when they are of the same `ArBasicKind`.  For instance, if you pass a `StructuredBuffer<Foo>` and `StructuredBuffer<Bar>`, will it correctly identify the mismatch?